pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,222 @@
+		failoverTimeout: time.Minute,
+		client:          client,
+		cancel:          cancel,
+	}
+	go t.loop(ctx)
+	return t, nil
+}
+
+// Follows pattern used by ring for WatchKey.
+func (c *tracker) loop(ctx context.Context) {
+	defer close(c.done)
+	// The KVStore config we gave when creating c should have contained a prefix,
+	// which would have given us a prefixed KVStore client. So, we can pass empty string here.
+	c.client.WatchPrefix(ctx, "", func(key string, value interface{}) bool {
+		if value == nil {
+			return true
+		}
+		replica := value.(*ReplicaDesc)
+		c.electedLock.Lock()
+		defer c.electedLock.Unlock()
+		c.elected[key] = *replica
+		return true
+	})
+}
+
+// Stop ends calls the trackers cancel function, which will end the loop for WatchPrefix.
+func (c *tracker) Stop() {
+	c.cancel()
+	<-c.done
+}
+
+// Wraps calls to client CAS with locking and error checking
+func (c *tracker) casWrapper(ctx context.Context, cluster, replica string, now time.Time) error {
+	// We have should a prefixed KVStore client, so we want to pass /cluster instead of just cluster.
+	// Prefixed client does prefix+key without separating the values.
+	err := c.client.CAS(ctx, fmt.Sprintf("/%s", cluster), func(in interface{}) (out interface{}, retry bool, err error) {
+		desc, ok := in.(*ReplicaDesc)
+		// There was either invalid or no data for the key, so we now accept samples
+		// from this replica.
+		if !ok || desc == nil {
+			return &ReplicaDesc{Replica: replica, ReceivedAt: timestamp.FromTime(now)}, false, nil
+		}
+
+		// As far as the timestamp in the KVStore is concerned none of the Distributors have
+		// received a sample from the stored replica recently, so we should overwrite the
+		// replica in the KVStore and accept this sample.
+		if now.Sub(timestamp.Time(desc.ReceivedAt)) > c.failoverTimeout {
+			desc.ReceivedAt = timestamp.FromTime(now)
+			desc.Replica = replica
+			return desc, false, nil
+		}
+
+		// Safeguard against other calls to the KVStore having changed the accepted replica via a
+		// CAS to the KVStore before us. We should only return a replica description when the
+		// replicas don't match in the case above, where the overwrite timeout has been reached.
+		if desc.Replica != replica {
+			// Return a 202.
+			return nil, false, httpgrpc.Errorf(http.StatusAccepted, "replicas did not match, rejecting sample: %s != %s", replica, desc.Replica)
+		}
+
+		// As far as the timestamp in the KVStore is concerned none of the Distributors have
+		// updated the timestamp for this cluster/replica in more than the write timeout.
+		// We should update the timestamp so that we don't potentially failover to accepting
+		// samples from another replica when we don't need to.
+		// This may only catch local CAS' updating the KVStore, not other Distributors.
+		if now.Sub(timestamp.Time(desc.ReceivedAt)) > c.updateTimeout {
+			desc.ReceivedAt = timestamp.FromTime(now)
+			return desc, false, nil
+		}
+		return nil, false, nil
```

Comment:
Stick a comment here saying `nil, false, nil` indicated success and will stop CAS overwriting the value.